Fails to parse existential type in statement position

As of rustc 1.29.0-nightly (874dec25e 2018-07-21) the following works:

```rust
#![feature(existential_type)]

use std::fmt::Debug;

existential type Existential: Debug;

fn main() {
    fn f() -> Existential {}
    println!("{:?}", f());
}
```

But the following does not parse. I would expect this to be equivalent to the above except for `Existential` in scope just within `main`, but if this is disallowed by the RFC then it should produce a better error message.

```rust
#![feature(existential_type)]

use std::fmt::Debug;

fn main() {
    existential type Existential: Debug;

    fn f() -> Existential {}
    println!("{:?}", f());
}
```

```console
error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `type`
 --> src/main.rs:6:17
  |
6 |     existential type Existential: Debug;
  |                 ^^^^ expected one of 8 possible tokens here

error[E0425]: cannot find value `existential` in this scope
 --> src/main.rs:6:5
  |
6 |     existential type Existential: Debug;
  |     ^^^^^^^^^^^ not found in this scope
```

Mentioning the existential types tracking issue https://github.com/rust-lang/rust/issues/34511
Mentioning @oli-obk and @cramertj

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -4391,6 +4391,11 @@ impl<'a> Parser<'a> {
         self.token.is_keyword(keywords::Extern) && self.look_ahead(1, |t| t != &token::ModSep)
     }
 
+    fn is_existential_type_decl(&self) -> bool {
+        self.token.is_keyword(keywords::Existential) &&
+        self.look_ahead(1, |t| t.is_keyword(keywords::Type))
+    }
+
     fn is_auto_trait_item(&mut self) -> bool {
         // auto trait
         (self.token.is_keyword(keywords::Auto)
@@ -4495,6 +4500,7 @@ impl<'a> Parser<'a> {
                   !self.is_union_item() &&
                   !self.is_crate_vis() &&
                   !self.is_extern_non_path() &&
+                  !self.is_existential_type_decl() &&
                   !self.is_auto_trait_item() {
             let pth = self.parse_path(PathStyle::Expr)?;
 

--- a/src/test/ui/impl-trait/existential_type_in_fn_body.rs
+++ b/src/test/ui/impl-trait/existential_type_in_fn_body.rs
@@ -1,0 +1,22 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// compile-pass
+
+#![feature(existential_type)]
+
+use std::fmt::Debug;
+
+fn main() {
+    existential type Existential: Debug;
+
+    fn f() -> Existential {}
+    println!("{:?}", f());
+}

